### ISSUE TITLE
Class::Tiny

### ISSUE THREAD

**Issue by tobyink** [memowe/EventStore-Tiny#10]
Have you considered using Class::Tiny instead of Mo?

Though the syntax is a little different, Class::Tiny classes can be subclassed more safely with Moo/Moose than Mo classes can. In particular, subclassing Mo classes with Moo or Moose can result in `BUILD` methods getting called multiple times.

**Comment by memowe** [comment]
Hey @tobyink, thank you very much for this suggestion!

Actually I didn't think a lot about the best choice as this was a tiny personal side project and I used Mo because I liked the minimal syntax. However, there are some parts (ingydotnet/mo-pm#36) I don't understand or that need to be fixed, so switching to another framework makes a lot of sense to me. I'll definitely have a more in-depth look at Class::Tiny as well as Mouse and Moo.

Do you know any recent comparisons?

**Comment by tobyink** [comment]
Yeah, Mo is a whole bag of crazy. All the [code golfing](https://raw.githubusercontent.com/ingydotnet/mo-pm/master/lib/Mo.pm) for example. It mostly works, but when it doesn't, it's a pain to figure out what went wrong.

Mouse, I wouldn't recommend. Although it's very similar to Moose, it's not quite similar enough for them to play nicely together. Trying to get a Mouse class to use a Moose role, for example, will give you headaches. It is fast though, because most of the important parts are written in XS. Though if you're using Perl somewhere with restrictive permissions, so you're limited to pure Perl modules, this is a bad thing.

Moo is pure Perl, but will take advantage of Class::XSAccessor to speed up accessors if available. Unlike Mouse, the constructors will never be XS though. It is slightly less Moose-like than Mouse on the surface, but goes to great effort to integrate well with Moose. Moo classes can consume Moose roles, and vice versa.

Class::Tiny is not quite as fast as Moo, but has zero dependencies. (I've written a patch for it to use Class::XSAccessor if it's installed like Moo does, but that hasn't been accepted yet.) The syntax is very non-Moose-like though.

```
package Person;
use Class::Tiny {
    "name"  => sub { die "name is required" },
    "age"   => sub { 0 }, # default
}, qw( email phone ); # optional attributes with no default
```

I think for what you're doing, Class::Tiny is sufficient. The only Mo features you're using are accessors with lazy defaults and a little bit of `required => 1`. And Class::Tiny supports those (see above example). It also supports `BUILD` which you're using. The only minor annoyance is that all Class::Tiny accessors are `rw` while you have a few `ro` ones. This can be easily worked around though. For example, if "name" is read-only:

```
package Person;
use Class::Tiny {
    "name"  => sub { die "name is required" },
    "age"   => sub { 0 }, # default
}, qw( email phone ); # optional attributes with no default

sub name { shift->{name} }   # manual ro accessor
```

If you write an accessor by hand, Class::Tiny won't try to overwrite it.

**Comment by memowe** [comment]
Thank you so much, @tobyink, for your help! I just merged my own PR #12: 388a4fd
